### PR TITLE
Implement prescribed mapping for pub date 

### LINKED ISSUE #571: fix pub year indexing
### Background
There is a desire to keep certain DOR/Argo Solr index fields as similar to SearchWorks indexing as possible.  This was previously handled by using the same code in the `stanford-mods` gem for both this app and for SearchWorks to get the values for Solr fields.

PR #517 decoupled dor indexing from Fedora (MODS), but was done without this realization, so the algorithms for retrieving values for Solr from our (Cocina) descriptive metadata were inadvertently over-simplified.  Also, the original tests for the dor_indexing_app were not at all thorough in exploring different data conditions and how they would be indexed.

### What Needs to be Done
This ticket is to make the `sw_pub_date_facet_ssi` field contain values as much like the ones we were getting from the stanford-mods method as is reasonable.

Be certain PR #553 is merged before embarking on this ticket - it addresses part of this.

Per PR #517, `sw_pub_date_facet_ssi` Solr field was formerly populated by calling `pub_year_display_str` in stanford-mods (https://github.com/sul-dlss/stanford-mods/blob/master/lib/stanford-mods/origin_info.rb#L47).  Tracing through the logic, you see it calls the `single_pub_year` method (https://github.com/sul-dlss/stanford-mods/blob/master/lib/stanford-mods/origin_info.rb#L188).  This does fun processing first on the MODS dateIssued elements looking for a pub date and calling year_display_str (https://github.com/sul-dlss/stanford-mods/blob/master/lib/stanford-mods/origin_info.rb#L82) if there is a dateIssued element, then, if there isn't one, does similar logic on dateCreated MODS elements, and then looks at dateCaptured MODS elements.  PR #553 more or less deals with what would have been the dateIssued elements, but not the other flavors of dates.   Even for finding only what were the dateIssued elements in the MODS, there is a 3-tiered logic to identify these values in Cocina.  

- Work with Arcadia to find the corresponding information ^^ in cocina models.
    - ask her if there are nuances to the data
- Write tests that exercise the different use cases
- ensure you account for the possibility of more complex data in the relevant cocina models by referring to the appropriate mapping specs in the dor-services-app, e.g.https://github.com/sul-dlss/dor-services-app/blob/main/spec/services/cocina/mapping/descriptive/mods/origin_info_date_spec.rb.  That is, if the code only expects "date.value" it won't find any values inside a structuredValue, a parallelValue (i.e. multi-lingual), and all that sort of stuff.

### User Impact

We are likely not populating the pub date field for some use cases and we are likely giving Argo users a mistaken impression of how their data will appear in SearchWorks.  This field is use in object titles in Argo, in a facet and probably in other places.

Note that in a perfect world, we will want to reindex all data that was indexed after PR #517 was deployed.

### PR DESCRIPTION
**Pull request by jcoyne**
## Why was this change made?
Fixes #571

### CODE REVIEW

File: spec/indexers/descriptive_metadata/pub_year_1e_spec.rb [hunk 1]
```diff
@@ -365,8 +365,8 @@
         }
       end
 
-      xit 'selects first date' do
-        expect(doc).to include('sw_pub_date_facet_ssi' => '')
+      it 'selects first date' do
+        expect(doc).to include('sw_pub_date_facet_ssi' => '2020')
```
Comment (jcoyne) [inline spec/indexers/descriptive_metadata/pub_year_1e_spec.rb:369]: @arcadiafalcone is this the correct value?

File: spec/indexers/descriptive_metadata/pub_year_1e_spec.rb [hunk 2]
```diff
@@ -570,8 +570,8 @@
         }
       end
 
-      xit 'selects date with status primary' do
-        expect(doc).to include('sw_pub_date_facet_ssi' => '2021-04-02')
+      it 'selects date with status primary' do
+        expect(doc).to include('sw_pub_date_facet_ssi' => '2022')
```
Comment (jcoyne) [inline spec/indexers/descriptive_metadata/pub_year_1e_spec.rb:574]: @arcadiafalcone this field is only an integer year, so I modified these few tests at the bottom.  Do they still look okay?

File: spec/indexers/descriptive_metadata/pub_year_1e_spec.rb (hunk 1 above)
Comment (arcadiafalcone) [reply spec/indexers/descriptive_metadata/pub_year_1e_spec.rb:369]: Yes.

File: spec/indexers/descriptive_metadata/pub_year_1e_spec.rb (hunk 2 above)
Comment (arcadiafalcone) [reply spec/indexers/descriptive_metadata/pub_year_1e_spec.rb:574]: Yes, these are all correct.

**Review by justinlittman — APPROVED** [review]

**Pull request merged**